pallets/parachain-system/src/lib.rs

```diff
@@ -544,6 +582,19 @@ pub mod pallet {
 		Unauthorized,
 	}
 
+	/// Latest parachain head data included in the relay chain.
+	///
+	/// This value is optional since it requires extra relay-chain state proof.
+	#[pallet::storage]
+	pub(super) type ParaHead<T: Config> = StorageValue<_, relay_chain::HeadData, OptionQuery>;
+
+	#[pallet::storage]
+	pub(super) type UnincludedSegment<T: Config> = StorageValue<_, Vec<BlockTracker>, ValueQuery>;
+
+	#[pallet::storage]
+	pub(super) type AggregatedUnincludedSegment<T: Config> =
+		StorageValue<_, SegmentTracker, OptionQuery>;
```

Comment:
Does the pallet require another runtime api exposed for appending into this queue?

Reply:
Appending should only happen within the runtime, so we shouldn't need a runtime API for that.

Reply:
i.e. `on_finalize` should do the work of appending to the unincluded segment and updating the aggregate tracker.

Reply:
I.e. this PR is missing the following in `on_finalize`:

If `MaxUnincludedLen` is present in the storage and `> 0`, then
construct `TotalBandwidthLimits` and try to append the block into unincluded segment.

?

Reply:
I'd expect some check in `on_initialize` which bails early if `hypothetical_segment_len_after_this_block > unincluded_segment_len` (the hypothetical is based on the pruning).

Then after `on_initialize` we'd always have the invariant that `unincluded_segment_len < max_unincluded_segment_len`, so we can unconditionally push the new bandwidth usage to the unincluded segment in `on_finalize`.

Reply:
I was thinking that we'd have the reduction in some kind of hook that runs before this check. Otherwise it does fail if the maximum gets reduced, true.

Reply:
Yeah, let's defer updating logic to another PR, so this one is just about the core mechanism